include/event_manager/event_manager_priv.h

```diff
@@ -0,0 +1,256 @@
+
+
+/* Declare a zero-length subscriber. */
+#define _EVENT_SUBSCRIBERS_EMPTY(ename, prio)										\
+	const struct {} _CONCAT(_EVENT_SUBSCRIBERS_SECTION_PREFIX(ename, prio), empty)	__used _EM_FORCED_ALIGNMENT	\
+	__attribute__((__section__(_EVENT_SUBSCRIBERS_SECTION_NAME(ename, prio)))) = {};
+
+
+/* Convenience macros generating section start and stop markers. */
+
+#define _EVENT_SUBSCRIBERS_START(ename, prio)	_CONCAT(__start_, _EVENT_SUBSCRIBERS_SECTION_PREFIX(ename, prio))
+
+#define _EVENT_SUBSCRIBERS_STOP(ename, prio)	_CONCAT(__stop_,  _EVENT_SUBSCRIBERS_SECTION_PREFIX(ename, prio))
+
+
+#define _EVENT_SUBSCRIBERS_DECLARE(ename)										\
+	extern const struct event_subscriber _EVENT_SUBSCRIBERS_START(ename, _SUBS_PRIO_ID(_SUBS_PRIO_FIRST))[];	\
+	extern const struct event_subscriber _EVENT_SUBSCRIBERS_STOP(ename,  _SUBS_PRIO_ID(_SUBS_PRIO_FIRST))[];	\
+	extern const struct event_subscriber _EVENT_SUBSCRIBERS_START(ename, _SUBS_PRIO_ID(_SUBS_PRIO_NORMAL))[];	\
+	extern const struct event_subscriber _EVENT_SUBSCRIBERS_STOP(ename,  _SUBS_PRIO_ID(_SUBS_PRIO_NORMAL))[];	\
+	extern const struct event_subscriber _EVENT_SUBSCRIBERS_START(ename, _SUBS_PRIO_ID(_SUBS_PRIO_FINAL))[];	\
+	extern const struct event_subscriber _EVENT_SUBSCRIBERS_STOP(ename,  _SUBS_PRIO_ID(_SUBS_PRIO_FINAL))[];
+
+
+/* Macro defining empty subscribers on each priority level.
+ * Each event type keeps an array of subscribers for every priority level.
+ * It can happen that for a given priority no subscriber will be registered.
+ * This macro declares zero-length subscriber that will cause required section
+ * to be generated by the linker. If no subscriber is registered at this
+ * level section will remain empty.
+ */
+#define _EVENT_SUBSCRIBERS_DEFINE(ename)					\
+	_EVENT_SUBSCRIBERS_EMPTY(ename, _SUBS_PRIO_ID(_SUBS_PRIO_FIRST))	\
+	_EVENT_SUBSCRIBERS_EMPTY(ename, _SUBS_PRIO_ID(_SUBS_PRIO_NORMAL))	\
+	_EVENT_SUBSCRIBERS_EMPTY(ename, _SUBS_PRIO_ID(_SUBS_PRIO_FINAL))
+
+
+/* Subscribe a listener to an event. */
+#define _EVENT_SUBSCRIBE(lname, ename, prio)										\
+	const struct event_subscriber _CONCAT(_CONCAT(__event_subscriber_, ename), lname) __used _EM_FORCED_ALIGNMENT	\
+	__attribute__((__section__(_EVENT_SUBSCRIBERS_SECTION_NAME(ename, prio)))) = {					\
+		.listener = &_CONCAT(__event_listener_, lname),								\
+	}
+
+
+
+/* Pointer to event type definition is used as event type identifier. */
+#define _EVENT_ID(ename) (&_CONCAT(__event_type_, ename))
+
+
+/* Macro generates a function of name new_ename where ename is provided as
+ * an argument. Allocator function is used to create an event of the given
+ * ename type.
+ */
+#define _EVENT_ALLOCATOR_FN(ename)					\
+	static inline struct ename *_CONCAT(new_, ename)(void)		\
+	{								\
+		struct ename *event = (struct ename *)k_malloc(sizeof(*event));\
+		BUILD_ASSERT(offsetof(struct ename, header) == 0,	\
+				 "");					\
+		if (unlikely(!event)) {					\
+			printk("Event Manager OOM error\n");		\
```

Comment:
i think this should be an option, typically when user uses dynamic allocation he need to take care about potential null return.
If you want to have failure here that should go through kernel api like `k_panic.

Reply:
agree, we should not be making decisions about reboots here. Also, using logging and not printk.

Reply:
the reason why logging was not used here was because it is a macro so module that is using it may have no logging defined. This can be workarounded by calling an internal event_manager function which will use logging from known logging space.